File: frontend/src/api/automation.ts
import client from './client'
import type { ListResponse } from './types'

export interface AutomationRule {
  id: number
  organization_id: number
  name: string
  description: string | null
  trigger_type: string
  trigger_config: Record<string, unknown> | null
  action_type: string
  action_config: Record<string, unknown> | null
  is_active: boolean
  execution_count: number
  last_executed_at: string | null
  created_at: string
  updated_at: string
}

export interface AutomationRuleCreate {
  name: string
  description?: string
  trigger_type: string
  trigger_config?: Record<string, unknown>
  action_type?: string
  action_config?: Record<string, unknown>
}

export interface AutomationRuleUpdate {
  name?: string
  description?: string
  trigger_type?: string
  trigger_config?: Record<string, unknown>
  action_type?: string
  action_config?: Record<string, unknown>
  is_active?: boolean
}

export interface AutomationLog {
  id: number
  rule_id: number
  trigger_entity_type: string
  trigger_entity_id: number | null
  action_result: Record<string, unknown> | null
  status: string
  error_message: string | null
  created_at: string
}

export const automationApi = {
  getList: async (params?: { skip?: number; limit?: number; trigger_type?: string }): Promise<ListResponse<AutomationRule>> => {
    const response = await client.get('/api/v1/automations/', { params })
    return response.data
  },

  get: async (id: number): Promise<AutomationRule> => {
    const response = await client.get(`/api/v1/automations/${id}`)
    return response.data
  },

  create: async (data: AutomationRuleCreate): Promise<AutomationRule> => {
    const response = await client.post('/api/v1/automations/', data)
    return response.data
  },

  update: async (id: number, data: AutomationRuleUpdate): Promise<AutomationRule> => {
    const response = await client.put(`/api/v1/automations/${id}`, data)
    return response.data
  },

  delete: async (id: number): Promise<void> => {
    await client.delete(`/api/v1/automations/${id}`)
  },

  toggle: async (id: number): Promise<AutomationRule> => {
    const response = await client.post(`/api/v1/automations/${id}/toggle`)
    return response.data
  },

  getLogs: async (params?: { rule_id?: number; skip?: number; limit?: number }): Promise<ListResponse<AutomationLog>> => {
    const response = await client.get('/api/v1/automations/logs/list', { params })
    return response.data
  },

  checkTimeTriggers: async (): Promise<void> => {
    await client.post('/api/v1/automations/check-time-triggers')
  },
}

File: backend/schemas/automation.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel

from backend.models.automation import AutomationTriggerType, AutomationActionType


class AutomationRuleCreate(BaseModel):
    name: str
    description: Optional[str] = None
    trigger_type: AutomationTriggerType
    trigger_config: Optional[dict] = None
    action_type: AutomationActionType = AutomationActionType.SEND_NOTIFICATION
    action_config: Optional[dict] = None


class AutomationRuleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    trigger_type: Optional[AutomationTriggerType] = None
    trigger_config: Optional[dict] = None
    action_type: Optional[AutomationActionType] = None
    action_config: Optional[dict] = None
    is_active: Optional[bool] = None


class AutomationRuleResponse(BaseModel):
    id: int
    organization_id: int
    name: str
    description: Optional[str] = None
    trigger_type: AutomationTriggerType
    trigger_config: Optional[dict] = None
    action_type: AutomationActionType
    action_config: Optional[dict] = None
    is_active: bool
    execution_count: int
    last_executed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AutomationLogResponse(BaseModel):
    id: int
    rule_id: int
    trigger_entity_type: str
    trigger_entity_id: Optional[int] = None
    action_result: Optional[dict] = None
    status: str
    error_message: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}
